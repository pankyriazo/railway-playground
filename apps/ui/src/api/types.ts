import { Deployments, Projects } from "@monorepo/types";

export type GetProjectListParams = {
  token: string;
};
export type GetProjectListResponse = Projects;

export type GetDeploymentListParams = {
  token: string;
  deploymentCount?: number;
  projectId: string;
  serviceId: string;
};
export type GetDeploymentListResponse = Deployments;
