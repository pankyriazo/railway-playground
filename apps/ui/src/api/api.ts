import { SERVER_URL } from "@/environment";

const baseUrl = SERVER_URL;

const api = {
  getProjectList: (token: string) => {
    return fetch(`${baseUrl}/projects?token=${token}`).then((res) =>
      res.json()
    );
  },
};

export default api;
