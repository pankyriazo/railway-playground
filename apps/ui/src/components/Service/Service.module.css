.skeleton {
  composes: skeleton from global;

  border-radius: var(--border-radius);
  height: 5rem;
  max-width: 100%;
  width: 10rem;
}

.title {
  color: var(--color-text--900);
  font-size: var(--font-size-md);
  font-weight: var(--font-weight-bold);
  margin-bottom: var(--spacing-sm);
}

.deployments {
  display: flex;
  flex-direction: column;
  row-gap: var(--spacing-xs);
}
