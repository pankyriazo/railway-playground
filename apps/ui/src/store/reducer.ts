import { Action, State } from "./types";
import { initialState } from "./store";

const reducer = (state: State, action: Action): State => {
  switch (action.type) {
    case "RESET":
      return initialState;
    case "SET_TOKEN":
      return { ...state, token: action.payload };
    case "SET_PROJECTS":
      return { ...state, projects: action.payload };
    case "ADD_DEPLOYMENT":
      return {
        ...state,
        deployments: { ...state.deployments, ...action.payload },
      };
    case "REMOVE_DEPLOYMENT":
      const { [action.payload]: _, ...deployments } = state.deployments;
      return { ...state, deployments };
    default:
      return state;
  }
};

export default reducer;
