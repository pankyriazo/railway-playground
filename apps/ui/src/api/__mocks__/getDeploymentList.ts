import { GetDeploymentListResponse } from "@/api/types";

export const mockGetDeploymentList = {
  data: {
    deployments: {
      edges: [
        {
          node: {
            id: "4b049de0-1ce0-4e81-8036-26b252e3285c",
            status: "SUCCESS",
            updatedAt: "2024-03-17T08:50:14.973Z",
            environmentId: "632da482-c5b1-4570-87af-9ea225c335cb",
          },
        },
        {
          node: {
            id: "38b68be4-f86d-4987-8c48-dee293fda404",
            status: "SUCCESS",
            updatedAt: "2024-03-16T19:41:25.747Z",
            environmentId: "1beb4373-6b1f-43cb-8150-0e7214acea82",
          },
        },
      ],
    },
  },
} as const satisfies GetDeploymentListResponse;

export const mockGetDeploymentListError = {
  data: null,
  errors: [
    {
      message: "Unauthorized",
    },
  ],
} as const satisfies GetDeploymentListResponse;
