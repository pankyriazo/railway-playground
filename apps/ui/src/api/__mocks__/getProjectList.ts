import { GetProjectListResponse } from "@/api/types";

export const mockGetProjectList = {
  data: {
    me: {
      projects: {
        edges: [
          {
            node: {
              id: "a20a0d44-a6de-4804-acf1-58d436570fd7",
              name: "nauseating-cup",
              services: {
                edges: [
                  {
                    node: {
                      id: "0e62e7bf-2008-4452-bd32-e57cec3bb9a2",
                      name: "fascinated-motion",
                    },
                  },
                  {
                    node: {
                      id: "7ab42902-aae8-47b7-9976-81496dc61834",
                      name: "Postgres",
                    },
                  },
                ],
              },
              plugins: {
                edges: [],
              },
              environments: {
                edges: [
                  {
                    node: {
                      id: "1beb4373-6b1f-43cb-8150-0e7214acea82",
                      name: "production",
                    },
                  },
                  {
                    node: {
                      id: "632da482-c5b1-4570-87af-9ea225c335cb",
                      name: "staging",
                    },
                  },
                ],
              },
            },
          },
        ],
      },
    },
  },
} as const satisfies GetProjectListResponse;

export const mockGetProjectListError = {
  data: null,
  errors: [
    {
      message: "Unauthorized",
    },
  ],
} as const satisfies GetProjectListResponse;
