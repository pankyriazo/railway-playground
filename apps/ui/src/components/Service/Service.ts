import { Deployments as DeploymentsType, Node } from "@monorepo/types";
import api from "@/api";
import { StatefulComponent } from "@/types";
import store from "@/store";
import Deployments from "@/components/Deployments";

import styles from "./Service.module.css";

const pendingFetchRequests = new Map<string, boolean>();

class Service extends StatefulComponent<HTMLLIElement> {
  private _token: string = "";
  private _project: Node;
  private _service: Node;
  private _environment: Node;
  private _deployments: NonNullable<DeploymentsType["data"]>["deployments"] = {
    edges: [],
  };

  constructor({
    project,
    service,
    environment,
  }: {
    project: Node;
    service: Node;
    environment: Node;
  }) {
    super();

    this._element = document.createElement("li");
    this._element.id = `service-${service.id}`;
    this._project = project;
    this._service = service;
    this._environment = environment;

    this._subscribe();

    this.render();
  }

  private _subscribe() {
    if (pendingFetchRequests.get(this._service.id)) {
      this._state = "loading";
    }

    return store.subscribe((state) => {
      if (this._service.id in state.deployments) {
        this._deployments = state.deployments[this._service.id]!;
        this.setState("loaded");

        return;
      }

      if (state.token !== this._token) {
        this._token = state.token;
        store.dispatch({
          type: "REMOVE_DEPLOYMENT",
          payload: this._service.id,
        });

        if (
          this._state !== "loading" &&
          !pendingFetchRequests.get(this._service.id)
        ) {
          this._fetchDeployments();
        }
      }
    });
  }

  private _fetchDeployments() {
    pendingFetchRequests.set(this._service.id, true);
    this.setState("loading");

    api
      .getDeploymentList({
        token: this._token,
        projectId: this._project.id,
        serviceId: this._service.id,
      })
      .then((response) => {
        pendingFetchRequests.delete(this._service.id);

        if (response.errors) {
          this.setState("error");
          return;
        }

        store.dispatch({
          type: "ADD_DEPLOYMENT",
          payload: {
            [this._service.id]: response.data.deployments,
          },
        });
      })
      .catch(() => {
        this.setState("error");
      });
  }

  override render() {
    if (!this._element) {
      return;
    }

    this._element.innerHTML = "";
    this._element.dataset.testid = "service";

    const title = document.createElement("h4");
    title.dataset.testid = "service__title";
    title.className = styles.title!;
    title.textContent = this._service.name;

    switch (this._state) {
      case "loading":
        const skeleton = document.createElement("div");
        skeleton.dataset.testid = "service__skeleton";
        skeleton.className = styles.skeleton!;
        this._element.appendChild(skeleton);
        break;
      case "loaded":
        if (!this._deployments) {
          throw new Error("Deployments not found");
        }

        this._element.appendChild(title);

        const deploymentListForEnvironment = this._deployments.edges.filter(
          ({ node: deployment }) =>
            deployment.environmentId === this._environment.id
        );

        const deploymentsInstance = new Deployments({
          deployments: { edges: deploymentListForEnvironment },
        });
        const deployments = document.createElement("div");
        deployments.className = styles.deployments!;
        deployments.appendChild(deploymentsInstance.element);
        this._element.appendChild(deployments);
        break;
      case "error":
        this._element.appendChild(title);

        const error = document.createElement("h5");
        error.dataset.testid = "service__error-title";
        error.textContent = "Error fetching deployments";
        this._element.appendChild(error);
        break;
    }
  }
}

export default Service;
