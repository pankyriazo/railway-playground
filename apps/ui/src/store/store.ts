import reducer from "./reducer";
import { Action, Listener, Reducer, State } from "./types";

export const initialState: State = {
  token: "",
  projects: {
    data: null,
    errors: [],
  },
  deployments: {},
};

class Store {
  private _state: State;
  private _reducer: Reducer;
  private _listeners: Listener[] = [];

  constructor() {
    this._state = initialState;
    this._reducer = reducer;
  }

  get state(): State {
    return this._state;
  }

  private _notify(): void {
    this._listeners.forEach((listener) => listener(this._state));
  }

  dispatch(action: Action): void {
    this._state = this._reducer(this._state, action);
    this._notify();
  }

  subscribe(listener: Listener): () => void {
    this._listeners.push(listener);

    listener(this._state);

    return () => {
      this._listeners = this._listeners.filter((l) => l !== listener);
    };
  }
}

const store = new Store();

export default store;
