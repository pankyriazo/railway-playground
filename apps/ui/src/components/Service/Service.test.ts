import { mockComponent, render } from "@/test-helpers";
import { mockGetProjectList } from "@/api/__mocks__/getProjectList";
import { mockGetDeploymentList } from "@/api/__mocks__/getDeploymentList";
import api from "@/api";
import store from "@/store";
import Service from "./Service";

jest.mock("@/components/Deployments", () => mockComponent("Deployments"));
jest.mock("@/api", () => ({
  getDeploymentList: () => Promise.resolve({} as any),
}));

describe("Service", () => {
  const project = mockGetProjectList.data.me.projects.edges[0].node;
  const mockProps = {
    project,
    service: project.services.edges[0].node,
    environment: project.environments.edges[0].node,
  };

  let apiSpy: jest.SpyInstance;

  beforeEach(() => {
    apiSpy = jest.spyOn(api, "getDeploymentList");

    render(Service, mockProps);
  });

  afterEach(() => {
    apiSpy.mockRestore();
  });

  describe("when the token is empty", () => {
    it("should not fetch deployments", () => {
      expect(apiSpy).not.toHaveBeenCalled();
    });

    it("should not render the title", () => {
      const title = document.querySelector("[data-testid=service__title]");

      expect(title).toBeNull();
    });

    it("should not render any deployments", () => {
      const deployments = document.querySelectorAll(
        "[data-component=Deployments]"
      );

      expect(deployments.length).toBe(0);
    });
  });

  describe("when the token is updated", () => {
    beforeEach(() => {
      store.dispatch({
        type: "SET_TOKEN",
        payload: "test-token",
      });
    });

    it("should fetch deployments", () => {
      expect(apiSpy).toHaveBeenCalled();
    });

    it("should render the title", () => {
      const title = document.querySelector("[data-testid=service__title]");

      expect(title!.textContent).toBe("fascinated-motion");
    });

    it("should render the loading state", async () => {
      const service = document.querySelector("[data-testid=service]");

      expect(service!.innerHTML).not.toBe("");
      expect(service).toMatchSnapshot();
    });
  });

  describe("when the deployments are loaded", () => {
    beforeEach(() => {
      store.dispatch({
        type: "ADD_DEPLOYMENT",
        payload: {
          [mockProps.service.id]: mockGetDeploymentList.data.deployments,
        },
      });
    });

    it("should render the title", () => {
      const title = document.querySelector("[data-testid=service__title]");

      expect(title!.textContent).toBe("fascinated-motion");
    });

    it("should render the deployments", () => {
      const service = document.querySelector("[data-testid=deployments]");

      const deployments = document.querySelectorAll(
        "[data-component=Deployments]"
      );

      expect(deployments.length).toBe(1);
      expect(service).toMatchSnapshot();
    });
  });
});
