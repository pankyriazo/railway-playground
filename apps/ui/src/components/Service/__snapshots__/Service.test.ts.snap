// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service when the deployments are loaded should render the deployments 1`] = `null`;

exports[`Service when the token is updated should render the loading state 1`] = `
<li
  data-testid="service"
  id="service-0e62e7bf-2008-4452-bd32-e57cec3bb9a2"
>
  <h4
    class="title"
    data-testid="service__title"
  >
    fascinated-motion
  </h4>
  <h5
    data-testid="service__error-title"
  >
    Error fetching deployments
  </h5>
</li>
`;
